calblueprint/airlock
# franco/server: Airlock Server

# PR Description
PR Description at a High Level

This PR accomplishes 4 main things

1. Creates a proxy server to redirect client requests from Airlock to Airtable
2. Provides /login, /register functionality through express
3.  Provides creation and verification of JWT tokens through asymmetric encryption
3.a Provides the option of of Blacklisting certain fields upon returning user payload `password` is blacklisted by default -- this is done in `createSafePayload`-- configuration done in the `XFIELDS` global variable in `lib/jwt.js`
3.b Provides the option of setting different expiration times + encryption scheme in `OPTIONS` variable in `lib/jwt.js`
4. Provides Hashing of passwords through Bcrypt and an optional flag to turn it off


Assumptions/Constraints Into PR

- Clients after logging-in/registering will need to provide a `token` in their header requests that contains the Airlock JWT token that was generated and provided after a client was registered/logged-in
- Developers using this server will need to ensure that they have a table within their Airtable base that contains the following two fields (exact case sensitivity)
    - `username`
    - `password`
- Developers will need to generate their own public/private rsa key


## Configuration Before Testing
- Create a Base in Airtable
- Obtain your Airtable api key
- Create a Table where you will store your users
    - Ensure this table has two fields `username` and `password` 
        - Both fields should be Single Line Text
- Fill out the `server/env.example` file in the root directory with the following environment variables — some have already been pre-populated (keep in mind if you want to test for plain-text passwords you will have to change `DISABLE_HASH_PASSWORD` to `true`)
    - `AIRTABLE_API_KEY`=YOUR_API_KEY
    - `AIRTABLE_BASE_ID`=YOUR_BASE_ID
    - `AIRTABLE_ENDPOINT_URL`=https://api.airtable.com/
    - `AIRTABLE_API_VERSION`=0.1.0
    - `AIRTABLE_USER_AGENT`=Airtable.js/0.7.1
    - `AIRTABLE_USER_TABLE`=NAME_OF_YOUR_USER_TABLE
    - `DISABLE_HASH_PASSWORD`=false
    - `SALT_ROUNDS`=PROVIDE_A_NUMBER
- Rename `env.example` to `.env`
- Fill out Public Key + Private Key
    - In the the `/server/lib` folder
    - Generate a set of public and private keys
    - Name these keys as `pub.pem` and `priv.pem`


## Running The Server
- If you have all of the configuration setup, go to the root directory of `/server` and type the following command
    - `npm run dev`
- If successful you should have a node server running on `http://localhost:4000/`


## How to Test
- Download postman (or any tool that you prefer that will allow you to simulate CRUD operations with the server)


## Testing Register
- Make a `POST` request to `http://localhost:4000/:version/:base/__DANGEROUSLY__USE__TABLE__TO__LET__USERS__REGISTER`
    - In the `BODY` provide a `username` and `password` in the form of JSON
        {"username": "franco",
        "password": "integrate"}
- Make the post request — if successful you should receive in return
     ```{
        "success": true,
        "token": "Jo__tAafsFuF6PU2YpbGWTolrlAjXAWNrudCaYnMrWm3RRxGz-31n721aKoWz8B7T4ofxHm_ZsxuZUoBGWm9_EeZDF_zrBnOrEij-tMi43DXOXByqskKeM5tR20Y8O8JeRygvdRYyayq5wqf5iOAxNxRINEgdB5HHLpG98B-sHQAHfs4UYPvDvCQT6XXHzEhbthYKht3Wmfg",
        "user": {
            "id": "recIOBFLPhcSwjRHQ",
            "fields": {
                "username": "franco"
            },
            "createdTime": "2019-10-14T02:40:45.000Z"
        }
    }



## Testing Login
- Make a `POST` request to `http://localhost:4000/:version/:base/__DANGEROUSLY__USE__TABLE__TO__LET__USERS__LOGIN` 
    - In the `BODY` provide the same `username` and `password` you used in testing `/register` in the form of JSON
- Make the post request — if successful you should receive in return
    ```{
        "success": true,
        "token": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InJlY2RVOVZEOU5acUIyWllDIiwiZmllbGRzIjp7InVzZXJuYW1lIjoiZnJhbmNvIn0sImNyZWF0ZWRUaW1lIjoiMjAxOS0xMC0xNFQwNDoyOTowNi4wMDBaIiwiaWF0IjoxNTcxMDI3MzYzfQ.bqB1EwGbx8-ud61aj_KDljLtYRQL1TS9o7b2e6_ATIQdNzgtFRt",
        "user": {
            "id": "recdU9VD9NZqB2ZYC",
            "fields": {
                "username": "franco"
            },
            "createdTime": "2019-10-14T04:29:06.000Z"
        }
    }


## Testing any request
- Background
    - In Airtable.js all queries come in the form of 
    var url = base._airtable._endpointUrl + '/v' + base._airtable._apiVersionMajor + '/' + base._id + path + '?' + objectToQueryParamString(queryParams);
    - Which translates to
        - `https://api.airtable.com/v0/BASE_ID/TABLE_NAME?fields%5B%5D=username&filterByFormula=username%3D%22franco%22`
- Goal
    - AirLock will take a request like `http://localhost:4000/v0/BASE_ID/TABLE_NAME?` which will be produced from the client
    - Airlock server will obtain this request, along with the Airlock token and then proxy the request by converting it into  `https://api.airtable.com/v0/BASAE_ID/TABLE_NAME?`
- **How to Test (you will notice some of these values come from the .env file)**
    - Open up Postman
    - Create a `GET` request
        - the url should be `http://localhost:4000/v0/BASE_ID/TABLE_NAME?`
        - this is preferably the table you are using for users
    - Configure the `Headers` tab to have the following headers and values
        - `x-airtable-application-id` = BASE_ID_VALUE
        - `x-airtable-user-agent` = Airtable.js/0.7.1
        - `x-api-version` = 0.1.0
        - `authorization` = Bearer YOUR_AIRTABLE_API_KEY
        - `token` = “**This value comes from the contents of the token you received from register/login”**
    - Now you can make a `GET` request
        - If successful you should be able to get all of the contents of the table
    
```{
        "records": [
            {
                "id": "recdU9VD9NZqB2ZYC",
                "fields": {
                    "username": "franco",
                    "password": "$2b$10$HxkyufAwdukUP90GG/SepC4tOfDl1UZViHm"
                },
                "createdTime": "2019-10-14T04:29:06.000Z"
            },
            {
                "id": "recjpfmEM4FasWzP3",
                "fields": {
                    "username": "arpan",
                    "password": "$2b$10$HxkyufAwdukUP90GG/epC4tOfDl1UZV"
                },
                "createdTime": "2019-10-13T23:45:13.000Z"
            }
        ]
    }
